### PR TITLE
8291003: ARM32: constant_table.size assertion

### PR DESCRIPTION
**Pull request by bulasevich**
This change fixes assertion condition as per the recent [JDK-8287373](https://bugs.openjdk.org/browse/JDK-8287373) change: the size of constants section is aligned up according to the settings of the next section (instructions section).

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8291003](https://bugs.openjdk.org/browse/JDK-8291003): ARM32: constant_table.size assertion

### Reviewers
 * [Aleksey Shipilev](https://openjdk.org/census#shade) (@shipilev - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/9672/head:pull/9672` \
`$ git checkout pull/9672`

Update a local copy of the PR: \
`$ git checkout pull/9672` \
`$ git pull https://git.openjdk.org/jdk pull/9672/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 9672`

View PR using the GUI difftool: \
`$ git pr show -t 9672`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/9672.diff">https://git.openjdk.org/jdk/pull/9672.diff</a>

</details>

### CODE REVIEW

**Review by shipilev — COMMENTED** [review]
This is equivalent to the following, right?

```
  int consts_size = consts_section->align_at_start(consts_section->size(), CodeBuffer::SECT_INSTS);
```

In fact, we can make it better by doing e.g.:

```
diff --git a/src/hotspot/cpu/arm/arm.ad b/src/hotspot/cpu/arm/arm.ad
index a99d8599cd3..c8e4b482778 100644
--- a/src/hotspot/cpu/arm/arm.ad
+++ b/src/hotspot/cpu/arm/arm.ad
@@ -236,7 +236,8 @@ void MachConstantBaseNode::emit(CodeBuffer& cbuf, PhaseRegAlloc* ra_) const {
 
   Register r = as_Register(ra_->get_encode(this));
   CodeSection* consts_section = __ code()->consts();
-  int consts_size = consts_section->align_at_start(consts_section->size());
+  // constants section size is aligned according to the align_at_start settings of the next section
+  int consts_size = CodeSection::align_at_start(consts_section->size(), CodeBuffer::SECT_INSTS);
   assert(constant_table.size() == consts_size, "must be: %d == %d", constant_table.size(), consts_size);
 
   // Materialize the constant table base.
diff --git a/src/hotspot/share/asm/codeBuffer.hpp b/src/hotspot/share/asm/codeBuffer.hpp
index e96f9c07e76..4cad3b4d30e 100644
--- a/src/hotspot/share/asm/codeBuffer.hpp
+++ b/src/hotspot/share/asm/codeBuffer.hpp
@@ -261,12 +261,12 @@ class CodeSection {
   // Slop between sections, used only when allocating temporary BufferBlob buffers.
   static csize_t end_slop()         { return MAX2((int)sizeof(jdouble), (int)CodeEntryAlignment); }
 
-  csize_t align_at_start(csize_t off, int section) const {
+  static csize_t align_at_start(csize_t off, int section) {
     return (csize_t) align_up(off, alignment(section));
   }
 
   csize_t align_at_start(csize_t off) const {
-    return (csize_t) align_up(off, alignment(_index));
+    return align_at_start(off, _index);
   }
 
   // Ensure there's enough space left in the current section.
```

**Comment by bulasevich** [comment]
> This is equivalent to the following, right?

Yes. I like your change in codeBuffer.hpp. Though your change in arm.ad brings a little bit too much details. And mine is shorter :) What do you think?
```
-  int consts_size = __ code()->insts()->align_at_start(consts_section->size());
+  int consts_size = CodeSection::align_at_start(consts_section->size(), CodeBuffer::SECT_INSTS);
```

**Review by shipilev — APPROVED** [review]
This is fine as well, but consider making it a bit cleaner, as discussed.

**Review by shipilev — APPROVED** [review]

**Comment by bulasevich** [comment]
I applied the changes to make it a bit cleaner. Thank you!

File: src/hotspot/cpu/arm/arm.ad [hunk 1]
```diff
@@ -236,7 +236,8 @@ void MachConstantBaseNode::emit(CodeBuffer& cbuf, PhaseRegAlloc* ra_) const {
 
   Register r = as_Register(ra_->get_encode(this));
   CodeSection* consts_section = __ code()->consts();
-  int consts_size = consts_section->align_at_start(consts_section->size());
```
Comment (TheFarlandsExplorer15) [inline src/hotspot/cpu/arm/arm.ad:239]: -     int consts_size = consts_section->align_at_start(consts_section->size());
I have no idea what that means and:
+    // constants section size is aligned according to the align_at_start settings of the next section
And it is: int consts_size = CodeSection::align_at_Start(consts_section->size(), CodeBuffer::SECT_INSTS);
Still no idea.

**Pull request closed without merging**